components/testing/COMPONENT_FPGA_CI_TEST_SHIELD/test_utils.h

```diff
@@ -0,0 +1,470 @@
+    } else {
+        test_peripheral<PortType, typename PortType::TestFunctionType, f>(matched_ports.front());
+    }
+}
+
+template <uint32_t N, typename PinMapType, typename FormFactorType, typename TestFunctionType>
+class Port;
+
+template <uint32_t N, typename PinMapType, typename FormFactorType, typename TestFunctionType>
+bool operator== (const Port<N, PinMapType, FormFactorType, TestFunctionType> &port1,
+                 const Port<N, PinMapType, FormFactorType, TestFunctionType> &port2);
+
+template <uint32_t N, typename PinMapType, typename FormFactorType, typename FunctionType>
+class Port {
+public:
+    int peripheral;
+    PinName pins[N];
+    PinName *ppins[N];
+
+    static const uint32_t pin_count = N;
+    typedef PinMapType PinMap;
+    typedef FunctionType TestFunctionType;
+
+    enum Status { StatusPass, StatusFail, StatusNotTested };
+    Status status;
+
+    Port(): peripheral(NC), status(StatusNotTested)
+    {
+        init_pins();
+    }
+
+    Port(const Port &port)
+    {
+        init_pins();
+        copy_from(port);
+    }
+
+    void init_pins()
+    {
+        for (uint32_t i = 0; i < N; i++) {
+            pins[i] = NC;
+            ppins[i] = &pins[i];
+        }
+    }
+
+    void copy_from(const Port &port)
+    {
+        peripheral = port.peripheral;
+        status = port.status;
+        for (uint32_t i = 0; i < N; i++) {
+            pins[i] = port.pins[i];
+        }
+    }
+
+    bool empty()
+    {
+        if (peripheral == NC) {
+            return true;
+        }
+        for (uint32_t i = 0; i < N; i++) {
+            if (pins[i] == NC) {
+                return true;
+            }
+        }
+        return false;
+    }
+
+    const char *str()
+    {
+        static char port_str[128];
+        char pin_str[32];
+        sprintf(port_str, "peripheral=(%d) ", peripheral);
+        for (uint32_t i = 0; i < N; i++) {
+            sprintf(pin_str, "%s=(%s) ", PinMap::pin_type_names[i], FormFactorType::pin_to_string(pins[i]));
+            strcat(port_str, pin_str);
+        }
+        return port_str;
+    }
+
+    friend bool operator==<> (const Port<N, PinMapType, FormFactorType, FunctionType> &port1, const Port<N, PinMapType, FormFactorType, FunctionType> &port2);
+};
+
+template <uint32_t N, typename PinMapType, typename FormFactorType, typename FunctionType>
+const uint32_t Port<N, PinMapType, FormFactorType, FunctionType>::pin_count;
+
+template <uint32_t N, typename PinMapType, typename FormFactorType, typename FunctionType>
+bool operator== (const Port<N, PinMapType, FormFactorType, FunctionType> &port1, const Port<N, PinMapType, FormFactorType, FunctionType> &port2)
+{
+    if (port1.peripheral != port2.peripheral) {
+        return false;
+    }
+    for (uint32_t i = 0; i < N; i++) {
+        if (port1.pins[i] != port2.pins[i]) {
+            return false;
+        }
+    }
+    return true;
+}
+
+class DefaultFormFactor {
```

Comment:
I've noticed these utils are not much documented in this header file - what this defaultformfactor does ?

Reply:
I added documentation to this.